File: go/manager.go
// manager.go — Task manager with goroutines, channels, and context
// Showcases: goroutines, channels, select, context, sync, generics (Go 1.21+)
package taskmanager

import (
	"context"
	"errors"
	"fmt"
	"sort"
	"sync"
)

// ── Manager interface ─────────────────────────────────────────────────────────

type Manager interface {
	Add(task *Task) error
	Get(id string) (*Task, bool)
	Remove(id string) bool
	All() []*Task
	Filter(pred func(*Task) bool) []*Task
}

// ── TaskManager ───────────────────────────────────────────────────────────────

type TaskManager struct {
	mu    sync.RWMutex
	store map[string]*Task
	name  string
}

func NewTaskManager(name string) *TaskManager {
	return &TaskManager{
		name:  name,
		store: make(map[string]*Task),
	}
}

// ── CRUD ──────────────────────────────────────────────────────────────────────

func (m *TaskManager) Add(task *Task) error {
	if task == nil {
		return errors.New("task must not be nil")
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	m.store[task.ID] = task
	return nil
}

func (m *TaskManager) Get(id string) (*Task, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	t, ok := m.store[id]
	return t, ok
}

func (m *TaskManager) Remove(id string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	_, ok := m.store[id]
	if ok {
		delete(m.store, id)
	}
	return ok
}

func (m *TaskManager) All() []*Task {
	m.mu.RLock()
	defer m.mu.RUnlock()
	tasks := make([]*Task, 0, len(m.store))
	for _, t := range m.store {
		tasks = append(tasks, t)
	}
	return tasks
}

func (m *TaskManager) Size() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.store)
}

// ── Query ─────────────────────────────────────────────────────────────────────

func (m *TaskManager) Filter(pred func(*Task) bool) []*Task {
	all := m.All()
	result := make([]*Task, 0)
	for _, t := range all {
		if pred(t) {
			result = append(result, t)
		}
	}
	return result
}

func (m *TaskManager) FindByStatus(statuses ...Status) []*Task {
	allowed := make(map[Status]struct{}, len(statuses))
	for _, s := range statuses {
		allowed[s] = struct{}{}
	}
	return m.Filter(func(t *Task) bool {
		_, ok := allowed[t.Status]
		return ok
	})
}

func (m *TaskManager) FindOverdue() []*Task {
	return m.Filter(func(t *Task) bool { return t.IsOverdue() })
}

func (m *TaskManager) RankByScore(topN int) []*Task {
	all := m.All()
	sort.Slice(all, func(i, j int) bool {
		return all[i].Score() > all[j].Score()
	})
	if topN > 0 && topN < len(all) {
		return all[:topN]
	}
	return all
}

func (m *TaskManager) GroupByStatus() map[Status][]*Task {
	groups := make(map[Status][]*Task)
	for _, t := range m.All() {
		groups[t.Status] = append(groups[t.Status], t)
	}
	return groups
}

// ── Concurrent processing ─────────────────────────────────────────────────────

// ProcessBatch runs a processing function over all tasks concurrently,
// collecting errors via a channel. Returns all errors encountered.
func (m *TaskManager) ProcessBatch(
	ctx context.Context,
	process func(ctx context.Context, t *Task) error,
	concurrency int,
) []error {
	tasks := m.All()
	sem := make(chan struct{}, concurrency)
	errCh := make(chan error, len(tasks))
	var wg sync.WaitGroup

	for _, task := range tasks {
		task := task // capture loop variable
		wg.Add(1)
		go func() {
			defer wg.Done()
			select {
			case <-ctx.Done():
				errCh <- ctx.Err()
				return
			case sem <- struct{}{}:
				defer func() { <-sem }()
			}
			if err := process(ctx, task); err != nil {
				errCh <- fmt.Errorf("task %s: %w", task.ID[:8], err)
			}
		}()
	}

	wg.Wait()
	close(errCh)

	var errs []error
	for e := range errCh {
		errs = append(errs, e)
	}
	return errs
}

// Watch streams tasks matching pred into a channel until ctx is cancelled.
func (m *TaskManager) Watch(ctx context.Context, pred func(*Task) bool) <-chan *Task {
	ch := make(chan *Task, 16)
	go func() {
		defer close(ch)
		for _, task := range m.Filter(pred) {
			select {
			case <-ctx.Done():
				return
			case ch <- task:
			}
		}
	}()
	return ch
}

// ── Bulk mutation ─────────────────────────────────────────────────────────────

func (m *TaskManager) BulkTransition(pred func(*Task) bool, newStatus Status) int {
	count := 0
	for _, task := range m.Filter(pred) {
		if err := task.Transition(newStatus); err == nil {
			count++
		}
	}
	return count
}

func (m *TaskManager) String() string {
	return fmt.Sprintf("TaskManager{name=%q, size=%d}", m.name, m.Size())
}

// ── Dead code ─────────────────────────────────────────────────────────────────

// legacyLinearSearch was replaced by map-based lookup. Never called.
func (m *TaskManager) legacyLinearSearch(id string) *Task {
	for _, t := range m.store {
		if t.ID == id {
			return t
		}
	}
	return nil
}

File: rust/manager.rs
// manager.rs — Generic task manager with iterators, closures, and trait objects
// Showcases: generics, trait bounds, iterators, closures, Box<dyn Trait>, HashMap
use std::collections::HashMap;

use crate::task::{Priority, Scorable, Status, Task, TaskError};

// ── Manager trait ─────────────────────────────────────────────────────────────

pub trait Manage<T> {
    fn add(&mut self, item: T);
    fn get(&self, id: &str) -> Option<&T>;
    fn remove(&mut self, id: &str) -> bool;
    fn all(&self) -> Vec<&T>;
    fn len(&self) -> usize;
    fn is_empty(&self) -> bool { self.len() == 0 }
}

// ── TaskManager ───────────────────────────────────────────────────────────────

pub struct TaskManager {
    name: String,
    store: HashMap<String, Task>,
}

impl TaskManager {
    pub fn new(name: impl Into<String>) -> Self {
        TaskManager {
            name: name.into(),
            store: HashMap::new(),
        }
    }

    // ── Queries using iterators + closures ────────────────────────────────

    pub fn filter<'a, F>(&'a self, pred: F) -> Vec<&'a Task>
    where
        F: Fn(&Task) -> bool,
    {
        self.store.values().filter(|t| pred(t)).collect()
    }

    pub fn find_by_status(&self, status: &Status) -> Vec<&Task> {
        self.filter(|t| &t.status == status)
    }

    pub fn find_overdue(&self) -> Vec<&Task> {
        self.filter(|t| t.is_overdue())
    }

    pub fn find_by_priority_at_least(&self, min: Priority) -> Vec<&Task> {
        self.filter(|t| t.priority >= min)
    }

    pub fn rank_by_score(&self, top_n: Option<usize>) -> Vec<&Task> {
        let mut tasks: Vec<&Task> = self.store.values().collect();
        tasks.sort_by(|a, b| b.score().partial_cmp(&a.score()).unwrap_or(std::cmp::Ordering::Equal));
        match top_n {
            Some(n) => tasks.into_iter().take(n).collect(),
            None    => tasks,
        }
    }

    pub fn group_by_status(&self) -> HashMap<String, Vec<&Task>> {
        let mut groups: HashMap<String, Vec<&Task>> = HashMap::new();
        for task in self.store.values() {
            groups.entry(task.status.to_string()).or_default().push(task);
        }
        groups
    }

    pub fn count_by_priority(&self) -> HashMap<String, usize> {
        let mut counts: HashMap<String, usize> = HashMap::new();
        for task in self.store.values() {
            *counts.entry(task.priority.to_string()).or_insert(0) += 1;
        }
        counts
    }

    // ── Aggregate stats ───────────────────────────────────────────────────

    pub fn completion_rate(&self) -> f64 {
        if self.store.is_empty() { return 0.0; }
        let done = self.store.values().filter(|t| t.status == Status::Done).count();
        done as f64 / self.store.len() as f64
    }

    pub fn average_effort(&self) -> f64 {
        if self.store.is_empty() { return 0.0; }
        let total: u32 = self.store.values().map(|t| t.estimate_effort()).sum();
        total as f64 / self.store.len() as f64
    }

    pub fn total_score(&self) -> f64 {
        self.store.values().map(|t| t.score()).sum()
    }

    // ── Bulk mutations ────────────────────────────────────────────────────

    pub fn bulk_transition<F>(&mut self, pred: F, new_status: Status) -> usize
    where
        F: Fn(&Task) -> bool,
    {
        let ids: Vec<String> = self.store
            .values()
            .filter(|t| pred(t))
            .map(|t| t.id.clone())
            .collect();

        let mut count = 0;
        for id in ids {
            if let Some(task) = self.store.get_mut(&id) {
                if task.transition(new_status.clone()).is_ok() {
                    count += 1;
                }
            }
        }
        count
    }

    pub fn remove_if<F>(&mut self, pred: F) -> usize
    where
        F: Fn(&Task) -> bool,
    {
        let before = self.store.len();
        self.store.retain(|_, t| !pred(t));
        before - self.store.len()
    }

    // ── Iterator adapters ─────────────────────────────────────────────────

    pub fn iter(&self) -> impl Iterator<Item = &Task> {
        self.store.values()
    }

    pub fn iter_active(&self) -> impl Iterator<Item = &Task> {
        self.store.values().filter(|t| !t.status.is_terminal())
    }

    /// Apply a transformation and collect into a Vec.
    pub fn map_collect<B, F>(&self, f: F) -> Vec<B>
    where
        F: Fn(&Task) -> B,
    {
        self.store.values().map(f).collect()
    }

    /// Fold over all tasks.
    pub fn fold<B, F>(&self, init: B, f: F) -> B
    where
        F: Fn(B, &Task) -> B,
    {
        self.store.values().fold(init, f)
    }
}

impl Manage<Task> for TaskManager {
    fn add(&mut self, item: Task) {
        self.store.insert(item.id.clone(), item);
    }

    fn get(&self, id: &str) -> Option<&Task> {
        self.store.get(id)
    }

    fn remove(&mut self, id: &str) -> bool {
        self.store.remove(id).is_some()
    }

    fn all(&self) -> Vec<&Task> {
        self.store.values().collect()
    }

    fn len(&self) -> usize {
        self.store.len()
    }
}

impl std::fmt::Display for TaskManager {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "TaskManager {{ name={:?}, size={}, completion={:.0}% }}",
            self.name, self.store.len(), self.completion_rate() * 100.0)
    }
}

// ── Dead code ─────────────────────────────────────────────────────────────────

/// Old linear search — replaced by HashMap lookup. Never called.
#[allow(dead_code)]
fn linear_search<'a>(store: &'a HashMap<String, Task>, id: &str) -> Option<&'a Task> {
    store.values().find(|t| t.id == id)
}
